src/core/unit.c

```diff
@@ -2308,44 +2292,69 @@ int unit_add_dependency(Unit *u, UnitDependency d, Unit *other, bool add_referen
                         return r;
         }
 
-        q = set_put(u->dependencies[d], other);
-        if (q < 0)
-                return q;
+        r = set_reserve(u->dependencies[d], 1);
+        if (r < 0)
+                return r;
 
-        if (inverse_table[d] != _UNIT_DEPENDENCY_INVALID && inverse_table[d] != d) {
-                v = set_put(other->dependencies[inverse_table[d]], u);
-                if (v < 0) {
-                        r = v;
-                        goto fail;
-                }
+        if (inverse_d != _UNIT_DEPENDENCY_INVALID && inverse_d != d) {
+                r = set_reserve(other->dependencies[inverse_d], 1);
+                if (r < 0)
+                        return r;
         }
 
         if (add_reference) {
-                w = set_put(u->dependencies[UNIT_REFERENCES], other);
-                if (w < 0) {
-                        r = w;
-                        goto fail;
-                }
+                r = set_reserve(u->dependencies[UNIT_REFERENCES], 1);
+                if (r < 0)
+                        return r;
 
-                r = set_put(other->dependencies[UNIT_REFERENCED_BY], u);
+                r = set_reserve(other->dependencies[UNIT_REFERENCED_BY], 1);
                 if (r < 0)
-                        goto fail;
+                        return r;
         }
-
-        unit_add_to_dbus_queue(u);
         return 0;
+}
 
-fail:
-        if (q > 0)
-                set_remove(u->dependencies[d], other);
+int unit_add_dependency(Unit *u, UnitDependency d, Unit *other, bool add_reference) {
 
-        if (v > 0)
-                set_remove(other->dependencies[inverse_table[d]], u);
+        Unit *orig_u = u, *orig_other = other;
+        int r;
+
+        assert(u);
+        assert(d >= 0 && d < _UNIT_DEPENDENCY_MAX);
+        assert(other);
 
-        if (w > 0)
-                set_remove(u->dependencies[UNIT_REFERENCES], other);
+        u = unit_follow_merge(u);
+        other = unit_follow_merge(other);
 
-        return r;
+        /* We won't allow dependencies on ourselves. We will not
+         * consider them an error however. */
+        if (u == other) {
+                maybe_warn_about_dependency(orig_u, orig_other->id, d);
+                return 0;
+        }
+
+        if (d == UNIT_BEFORE && other->type == UNIT_DEVICE) {
+                log_unit_warning(u, "Dependency Before=%s ignored (.device units cannot be delayed)", other->id);
+                return 0;
+        }
+
+        r = unit_reserve_dependency(u, d, other, add_reference);
+        if (r < 0)
+                return r;
+
+        /* The following operations can't fail. */
+        assert(set_put(u->dependencies[d], other) >= 0);
```

Comment:
no we currently cannot invalidate deps. But I am fine with leaving the deps in place on failure. Failure is failure after all, things won't work in that case anyway...